File: day-04/logic/logic.py
from util.types import Board

MARK = -1


def mark_number_if_present(number: int, board: Board):
    for i in range(len(board)):
        for j in range(len(board[i])):
            if board[i][j] == number:
                board[i][j] = MARK


def has_winning_row(board: Board) -> bool:
    for row in board:
        if all(x == MARK for x in row):
            return True

    for j in range(len(board[0])):
        if all(board[i][j] == MARK for i in range(len(board))):
            return True

    return False


def sum_of_unmarked_numbers(board: Board) -> int:
    sum = 0
    for row in board:
        for cell in row:
            if cell != MARK:
                sum += cell

    return sum

File: day-04/logic/logic.go
package logic

import "github.com/dawidl022/advent-of-code-2021/day-04/util"

const MARK = -1

func MarkNumberIfPresent(number int, board *util.Board) {
	b := board.Rows

	for i := 0; i < len(b); i++ {
		for j := 0; j < len(b[i]); j++ {
			if b[i][j] == number {
				b[i][j] = MARK
			}
		}
	}
	board.Rows = b
}

func HasWinningRow(board *util.Board) bool {
	b := board.Rows

	for _, row := range b {
		allMarked := true
		for _, cell := range row {
			if cell != MARK {
				allMarked = false
				break
			}
		}
		if allMarked {
			return true
		}
	}

	for j := 0; j < len(b[0]); j++ {
		allMarked := true
		for i := 0; i < len(b); i++ {
			if b[i][j] != MARK {
				allMarked = false
			}
		}
		if allMarked {
			return true
		}
	}
	return false
}

func SumOfUnmarkedNumbers(board *util.Board) int {
	sum := 0

	for _, row := range board.Rows {
		for _, cell := range row {
			if cell != MARK {
				sum += cell
			}
		}
	}
	return sum
}
